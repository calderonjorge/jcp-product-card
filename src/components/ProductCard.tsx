import React , { createContext, useContext, CSSProperties } from 'react';
import { useProduct } from '../hooks/useProduct';

import { ProductContextProps, Product, onChangeArgs, InitialValues, ProductCardHandlers } from '../interfaces/Product.interfaces';
import styles from '../styles/styles.module.css';

const ProductContext = createContext<ProductContextProps>({} as ProductContextProps);

export const useProductContext = () => useContext(ProductContext);

export interface Props {
    product: Product,
    children: (args: ProductCardHandlers) => JSX.Element,
    initialValues?: InitialValues,
    className?: string,
    style?: CSSProperties,
    value?: number,
    onChange?: (args: onChangeArgs) => void,
}

export const ProductCard = ({ product, children, className, style, value, onChange, initialValues }: Props) => {
    
    const { counter, increaseBy, maxCount, isMaxCountReached, reset } = useProduct({ onChange, product, value, initialValues });


    return (
        <ProductContext.Provider value={{ counter, increaseBy, product, maxCount, isMaxCountReached }}>
            <div className={`${styles.productCard} ${className}`} style={style}>
                {children(
                    {
                        count: counter,
                        increaseBy,
                        isMaxCountReached,
                        maxCount,
                        product,
                        reset,
                    }
                )}
            </div>
        </ProductContext.Provider>
    )
}
