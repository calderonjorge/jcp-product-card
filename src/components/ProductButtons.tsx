import React from 'react';
import { useProductContext } from "./ProductCard";
import styles from '../styles/styles.module.css';
import { CSSProperties } from "react";

export interface Props {
    maxCount?: number;
    className?: string;
    style?: CSSProperties;
}

export const ProductButtons = ({ className, style }: Props) => {

    const { increaseBy, counter, isMaxCountReached } = useProductContext();
    
    return (
        <div className={`${styles.buttonsContainer} ${className}`} style={style}>
            <button className={styles.buttonMinus} onClick={() => increaseBy(-1)}>-</button>
            <div className={styles.countLabel}>{counter}</div>
            <button
                className={`${styles.buttonAdd} ${isMaxCountReached && styles.disabled}`}
                onClick={() => { increaseBy(1) }}
                disabled={isMaxCountReached}
            >
                +
            </button>
        </div>
    )
}