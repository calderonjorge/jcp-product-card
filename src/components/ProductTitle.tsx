import React from 'react';
import { useProductContext } from "./ProductCard";
import styles from '../styles/styles.module.css'
import { CSSProperties } from "react";

export interface Props {
    className?: string
    title?: string,
    style?: CSSProperties
}

export const ProductTitle = ({ title, className, style }: Props) => {
    const { product } = useProductContext();

    return <span className={`${styles.productDescription} ${className}`} style={style}>{title ? title : product.title}</span>
}