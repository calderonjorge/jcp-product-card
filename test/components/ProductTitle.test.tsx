import React from 'react';
import rendered from 'react-test-renderer'
import { ProductTitle, ProductCard } from '../../src/components';
import { product } from '../data/products';

describe('Product Title', () => {
    test('should show a custom title', () => {
        const wraper = rendered.create(<ProductTitle title='Custom Product' className='custom class' />);
        expect(wraper.toJSON()).toMatchSnapshot();
    });

    test('should show a component with product name', () => {
        const wraper = rendered.create(
            <ProductCard product={product}>
                {
                    () => (
                        <ProductTitle />
                    )
                }
            </ProductCard>
        );
        expect(wraper.toJSON()).toMatchSnapshot();
    });
})
