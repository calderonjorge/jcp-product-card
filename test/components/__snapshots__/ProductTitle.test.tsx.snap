// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Title should show a component with product name 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
    Coffee Mug - Without Image
  </span>
</div>
`;

exports[`Product Title should show a custom title 1`] = `
<span
  className="productDescription custom class"
>
  Custom Product
</span>
`;
